File: backend/utils/market_urls.py
from __future__ import annotations

import re
from collections.abc import Mapping, MutableMapping
from typing import Any
from urllib.parse import quote

POLYMARKET_BASE_URL = "https://polymarket.com"
KALSHI_BASE_URL = "https://kalshi.com"

_CONDITION_ID_RE = re.compile(r"^0x[0-9a-fA-F]+$")
_KALSHI_TICKER_RE = re.compile(r"^KX[A-Z0-9-]+$")
_POLYMARKET_SLUG_RE = re.compile(r"^(?=.*[a-z])[a-z0-9-]+$")
_KALSHI_OUTCOME_SEGMENT_RE = re.compile(r"^[A-Z0-9]{1,5}$")


def _clean_text(value: Any) -> str:
    return str(value or "").strip()


def _clean_segment(value: Any) -> str:
    return _clean_text(value).strip("/")


def _is_condition_id(value: str) -> bool:
    return bool(_CONDITION_ID_RE.fullmatch(value))


def _is_likely_kalshi_ticker(value: Any) -> bool:
    ticker = _clean_segment(value).upper()
    return bool(ticker and _KALSHI_TICKER_RE.fullmatch(ticker))


def _is_likely_polymarket_slug(value: Any) -> bool:
    slug = _clean_segment(value).lower()
    return bool(slug and _POLYMARKET_SLUG_RE.fullmatch(slug))


def _normalize_kalshi_ticker(value: Any) -> str:
    return re.sub(r"_(yes|no)$", "", _clean_segment(value), flags=re.IGNORECASE)


def derive_kalshi_event_ticker(market_ticker: Any) -> str:
    ticker = _normalize_kalshi_ticker(market_ticker).upper()
    if not ticker:
        return ""

    # Kalshi market tickers for multi-outcome events often append a short
    # outcome segment to the event ticker (e.g. "...-MOR"). Strip only that
    # trailing segment when it looks like an outcome code.
    parts = [p for p in ticker.split("-") if p]
    if len(parts) >= 3:
        last = parts[-1]
        if _KALSHI_OUTCOME_SEGMENT_RE.fullmatch(last):
            candidate = "-".join(parts[:-1])
            if _is_likely_kalshi_ticker(candidate):
                return candidate
    return ticker


def derive_kalshi_series_ticker(value: Any) -> str:
    ticker = _normalize_kalshi_ticker(value).upper()
    if not ticker:
        return ""
    parts = [p for p in ticker.split("-") if p]
    if not parts:
        return ""
    return parts[0]


def infer_market_platform(market: Mapping[str, Any]) -> str:
    explicit = _clean_segment(market.get("platform")).lower()
    if explicit == "kalshi":
        return "kalshi"
    if explicit == "polymarket":
        return "polymarket"

    condition_id = _clean_segment(market.get("condition_id") or market.get("conditionId"))
    if _is_condition_id(condition_id):
        return "polymarket"

    market_id = _clean_segment(market.get("id") or market.get("market_id") or market.get("ticker"))
    market_slug = _clean_segment(market.get("slug") or market.get("market_slug"))
    event_ticker = _clean_segment(market.get("event_ticker") or market.get("eventTicker"))
    if (
        _is_likely_kalshi_ticker(market_id)
        or _is_likely_kalshi_ticker(market_slug)
        or _is_likely_kalshi_ticker(event_ticker)
    ):
        return "kalshi"

    return "polymarket"


def build_polymarket_market_url(
    *,
    event_slug: Any = None,
    market_slug: Any = None,
    market_id: Any = None,
    condition_id: Any = None,
) -> str | None:
    event_slug_text = _clean_segment(event_slug).lower()
    market_slug_text = _clean_segment(market_slug).lower()
    market_id_text = _clean_segment(market_id).lower()
    condition_id_text = _clean_segment(condition_id).lower()

    if event_slug_text and market_slug_text and event_slug_text != market_slug_text:
        return f"{POLYMARKET_BASE_URL}/event/{quote(event_slug_text, safe='')}/{quote(market_slug_text, safe='')}"
    if market_slug_text:
        # /market/{market_slug} is stable and resolves to the canonical event path.
        return f"{POLYMARKET_BASE_URL}/market/{quote(market_slug_text, safe='')}"
    if event_slug_text:
        return f"{POLYMARKET_BASE_URL}/event/{quote(event_slug_text, safe='')}"

    # Never use numeric/condition IDs for website URLs; they 404 on polymarket.com.
    if not _is_condition_id(condition_id_text) and _is_likely_polymarket_slug(market_id_text):
        return f"{POLYMARKET_BASE_URL}/market/{quote(market_id_text, safe='')}"
    return None


def build_kalshi_market_url(
    *,
    market_ticker: Any = None,
    event_ticker: Any = None,
    event_slug: Any = None,
    series_ticker: Any = None,
) -> str | None:
    # Kalshi market pages are canonically addressed as:
    # /markets/{series_ticker}/{event_ticker}
    # (which redirects to /markets/{series}/{human-readable-title}/{event}).
    explicit_event = _clean_segment(event_ticker) or _clean_segment(event_slug)
    if _is_likely_kalshi_ticker(explicit_event):
        event_ticker_text = explicit_event.upper()
    else:
        event_ticker_text = derive_kalshi_event_ticker(market_ticker)

    explicit_series = _clean_segment(series_ticker)
    if _is_likely_kalshi_ticker(explicit_series):
        series_ticker_text = explicit_series.upper()
    else:
        series_ticker_text = derive_kalshi_series_ticker(event_ticker_text or market_ticker)

    if _is_likely_kalshi_ticker(event_ticker_text):
        if _is_likely_kalshi_ticker(series_ticker_text) and series_ticker_text != event_ticker_text:
            return (
                f"{KALSHI_BASE_URL}/markets/"
                f"{quote(series_ticker_text.lower(), safe='')}/"
                f"{quote(event_ticker_text.lower(), safe='')}"
            )
        return f"{KALSHI_BASE_URL}/markets/{quote(event_ticker_text.lower(), safe='')}"

    return None


def build_market_url(
    market: Mapping[str, Any],
    *,
    opportunity_event_slug: Any = None,
) -> str | None:
    platform = infer_market_platform(market)
    if platform == "kalshi":
        # For Kalshi, "slug" / "market_slug" contain the full market ticker
        # (e.g. "KXPOLITICSMENTION-26FEB15-SHUT"), NOT the event slug.
        # Only pass explicitly event-scoped fields to avoid building URLs
        # from the full market ticker.
        return build_kalshi_market_url(
            market_ticker=market.get("id") or market.get("market_id") or market.get("ticker"),
            event_ticker=(market.get("event_ticker") or market.get("eventTicker")),
            event_slug=(market.get("event_slug") or market.get("eventSlug")),
            series_ticker=(market.get("series_ticker") or market.get("seriesTicker")),
        )

    return build_polymarket_market_url(
        event_slug=market.get("event_slug") or market.get("eventSlug") or opportunity_event_slug,
        market_slug=market.get("slug") or market.get("market_slug"),
        market_id=market.get("id") or market.get("market_id"),
        condition_id=market.get("condition_id") or market.get("conditionId"),
    )


def _clean_absolute_url(value: Any) -> str:
    text = _clean_text(value)
    if text.startswith("http://") or text.startswith("https://"):
        return text
    return ""


def attach_market_links_to_opportunity_dict(
    opportunity: MutableMapping[str, Any] | Mapping[str, Any],
) -> dict[str, Any]:
    data = dict(opportunity)
    event_slug = _clean_segment(data.get("event_slug") or data.get("eventSlug"))

    polymarket_url = _clean_absolute_url(data.get("polymarket_url"))
    kalshi_url = _clean_absolute_url(data.get("kalshi_url"))

    market_rows = data.get("markets")
    markets = market_rows if isinstance(market_rows, list) else []
    enriched_markets: list[Any] = []

    for market_entry in markets:
        if not isinstance(market_entry, Mapping):
            enriched_markets.append(market_entry)
            continue

        market_dict = dict(market_entry)
        platform = infer_market_platform(market_dict)
        market_dict["platform"] = platform

        market_url = _clean_absolute_url(market_dict.get("url") or market_dict.get("market_url"))
        if not market_url:
            market_url = build_market_url(market_dict, opportunity_event_slug=event_slug)

        if market_url:
            market_dict["url"] = market_url
            market_dict["market_url"] = market_url
            if platform == "polymarket" and not polymarket_url:
                polymarket_url = market_url
            if platform == "kalshi" and not kalshi_url:
                kalshi_url = market_url

        enriched_markets.append(market_dict)

    data["markets"] = enriched_markets
    data["polymarket_url"] = polymarket_url or None
    data["kalshi_url"] = kalshi_url or None
    return data


def serialize_opportunity_with_links(opportunity: Any) -> dict[str, Any]:
    if isinstance(opportunity, Mapping):
        payload = dict(opportunity)
    elif hasattr(opportunity, "model_dump"):
        payload = opportunity.model_dump(mode="json")
    else:
        raise TypeError("Unsupported opportunity payload type")
    return attach_market_links_to_opportunity_dict(payload)

File: frontend/src/lib/marketUrls.ts
type NullableString = string | null | undefined
export type MarketPlatform = 'polymarket' | 'kalshi'

const POLYMARKET_BASE_URL = 'https://polymarket.com'
const KALSHI_BASE_URL = 'https://kalshi.com'

function cleanSegment(value: NullableString): string {
  return (value || '').trim().replace(/^\/+|\/+$/g, '')
}

function encodeSegment(value: string): string {
  return encodeURIComponent(value)
}

function isConditionId(value: string): boolean {
  return /^0x[0-9a-fA-F]+$/.test(value)
}

function isLikelyKalshiTicker(value: NullableString): boolean {
  const ticker = cleanSegment(value).toUpperCase()
  return /^KX[A-Z0-9-]+$/.test(ticker)
}

function isLikelyKalshiOutcomeSegment(value: string): boolean {
  return /^[A-Z0-9]{1,5}$/.test(value)
}

function isLikelyPolymarketSlug(value: NullableString): boolean {
  const slug = cleanSegment(value).toLowerCase()
  return /^(?=.*[a-z])[a-z0-9-]+$/.test(slug)
}

function normalizeKalshiTicker(value: NullableString): string {
  return cleanSegment(value).replace(/_(yes|no)$/i, '')
}

export function deriveKalshiEventTicker(marketTicker: NullableString): string {
  const ticker = normalizeKalshiTicker(marketTicker).toUpperCase()
  if (!ticker) return ''

  // Multi-outcome market tickers often append a short outcome code segment
  // to the event ticker (e.g. "...-MOR"). Strip that segment when present.
  const parts = ticker.split('-').filter(Boolean)
  if (parts.length >= 3) {
    const maybeOutcome = parts[parts.length - 1]
    if (isLikelyKalshiOutcomeSegment(maybeOutcome)) {
      const candidate = parts.slice(0, -1).join('-')
      if (isLikelyKalshiTicker(candidate)) return candidate
    }
  }
  return ticker
}

export function deriveKalshiSeriesTicker(value: NullableString): string {
  const ticker = normalizeKalshiTicker(value).toUpperCase()
  if (!ticker) return ''
  const parts = ticker.split('-').filter(Boolean)
  if (!parts.length) return ''
  return parts[0]
}

export function inferMarketPlatform(params: {
  platform?: NullableString
  marketId?: NullableString
  marketSlug?: NullableString
  conditionId?: NullableString
  eventTicker?: NullableString
}): MarketPlatform {
  const explicit = cleanSegment(params.platform).toLowerCase()
  if (explicit === 'kalshi') return 'kalshi'
  if (explicit === 'polymarket') return 'polymarket'

  const conditionId = cleanSegment(params.conditionId)
  if (isConditionId(conditionId)) return 'polymarket'

  if (
    isLikelyKalshiTicker(params.marketId)
    || isLikelyKalshiTicker(params.marketSlug)
    || isLikelyKalshiTicker(params.eventTicker)
  ) {
    return 'kalshi'
  }

  return 'polymarket'
}

export function buildPolymarketMarketUrl(params: {
  eventSlug?: NullableString
  marketSlug?: NullableString
  marketId?: NullableString
  conditionId?: NullableString
}): string | null {
  const eventSlug = cleanSegment(params.eventSlug).toLowerCase()
  const marketSlug = cleanSegment(params.marketSlug).toLowerCase()
  const marketId = cleanSegment(params.marketId).toLowerCase()
  const conditionId = cleanSegment(params.conditionId)

  if (eventSlug && marketSlug && eventSlug !== marketSlug) {
    return `${POLYMARKET_BASE_URL}/event/${encodeSegment(eventSlug)}/${encodeSegment(marketSlug)}`
  }
  if (marketSlug) {
    // /market/{market_slug} redirects to canonical event routes.
    return `${POLYMARKET_BASE_URL}/market/${encodeSegment(marketSlug)}`
  }
  if (eventSlug) {
    return `${POLYMARKET_BASE_URL}/event/${encodeSegment(eventSlug)}`
  }
  // condition/token/numeric IDs are not stable website routes on polymarket.com.
  if (!isConditionId(conditionId) && isLikelyPolymarketSlug(marketId) && !isLikelyKalshiTicker(marketId)) {
    return `${POLYMARKET_BASE_URL}/market/${encodeSegment(marketId)}`
  }
  return null
}

export function buildKalshiMarketUrl(params: {
  marketTicker?: NullableString
  eventTicker?: NullableString
  eventSlug?: NullableString
  seriesTicker?: NullableString
}): string | null {
  // Kalshi market pages resolve via:
  // /markets/{series_ticker}/{event_ticker}
  // which redirects to the full canonical path with title slug.
  const explicitEvent = cleanSegment(params.eventTicker) || cleanSegment(params.eventSlug)
  const eventTicker = isLikelyKalshiTicker(explicitEvent)
    ? explicitEvent.toUpperCase()
    : deriveKalshiEventTicker(params.marketTicker)

  const explicitSeries = cleanSegment(params.seriesTicker)
  const seriesTicker = isLikelyKalshiTicker(explicitSeries)
    ? explicitSeries.toUpperCase()
    : deriveKalshiSeriesTicker(eventTicker || params.marketTicker)

  if (isLikelyKalshiTicker(eventTicker)) {
    if (isLikelyKalshiTicker(seriesTicker) && seriesTicker !== eventTicker) {
      return `${KALSHI_BASE_URL}/markets/${encodeSegment(seriesTicker.toLowerCase())}/${encodeSegment(eventTicker.toLowerCase())}`
    }
    return `${KALSHI_BASE_URL}/markets/${encodeSegment(eventTicker.toLowerCase())}`
  }

  return null
}

function cleanAbsoluteUrl(value: NullableString): string | null {
  const text = (value || '').trim()
  if (!text) return null
  if (text.startsWith('http://') || text.startsWith('https://')) return text
  return null
}

type OpportunityMarketForLinks = {
  id?: NullableString
  market_id?: NullableString
  ticker?: NullableString
  slug?: NullableString
  market_slug?: NullableString
  event_slug?: NullableString
  event_ticker?: NullableString
  series_ticker?: NullableString
  seriesTicker?: NullableString
  condition_id?: NullableString
  conditionId?: NullableString
  platform?: NullableString
  url?: NullableString
  market_url?: NullableString
}

type OpportunityForLinks = {
  event_slug?: NullableString
  markets?: OpportunityMarketForLinks[]
  polymarket_url?: NullableString
  kalshi_url?: NullableString
}

export type OpportunityLinkEntry = {
  platform: MarketPlatform
  url: string | null
}

export type OpportunityPlatformLinks = {
  polymarketUrl: string | null
  kalshiUrl: string | null
  marketLinks: OpportunityLinkEntry[]
}

// Single global resolver used by opportunities views. It prefers API-provided links.
export function getOpportunityPlatformLinks(opportunity: OpportunityForLinks | null | undefined): OpportunityPlatformLinks {
  const markets = Array.isArray(opportunity?.markets) ? opportunity!.markets : []
  const eventSlug = opportunity?.event_slug

  let polymarketUrl = cleanAbsoluteUrl(opportunity?.polymarket_url)
  let kalshiUrl = cleanAbsoluteUrl(opportunity?.kalshi_url)

  const marketLinks = markets.map((market): OpportunityLinkEntry => {
    const platform = inferMarketPlatform({
      platform: market.platform,
      marketId: market.id || market.market_id || market.ticker,
      marketSlug: market.slug || market.market_slug,
      conditionId: market.condition_id || market.conditionId,
      eventTicker: market.event_ticker,
    })

    const apiUrl = cleanAbsoluteUrl(market.url || market.market_url)
    const fallbackUrl = platform === 'kalshi'
      ? buildKalshiMarketUrl({
          marketTicker: market.id || market.market_id || market.ticker,
          eventTicker: market.event_ticker,
          eventSlug: market.event_slug,
          seriesTicker: market.series_ticker || market.seriesTicker,
        })
      : buildPolymarketMarketUrl({
          eventSlug: market.event_slug || eventSlug,
          marketSlug: market.slug || market.market_slug,
          marketId: market.id || market.market_id,
          conditionId: market.condition_id || market.conditionId,
        })

    const url = apiUrl || fallbackUrl || null
    if (platform === 'polymarket' && !polymarketUrl && url) polymarketUrl = url
    if (platform === 'kalshi' && !kalshiUrl && url) kalshiUrl = url

    return { platform, url }
  })

  return { polymarketUrl, kalshiUrl, marketLinks }
}

type TraderOrderLinkInput = {
  source?: NullableString
  marketId?: NullableString
  marketQuestion?: NullableString
  payload?: Record<string, unknown> | null | undefined
}

function toRecord(value: unknown): Record<string, unknown> {
  if (value && typeof value === 'object' && !Array.isArray(value)) {
    return value as Record<string, unknown>
  }
  return {}
}

function normalizeSourcePlatform(source: NullableString): MarketPlatform | null {
  const key = cleanSegment(source).toLowerCase()
  if (!key) return null
  if (key.includes('kalshi')) return 'kalshi'
  if (key.includes('poly') || key.includes('crypto')) return 'polymarket'
  return null
}

function buildPolymarketSearchUrl(query: NullableString): string | null {
  const text = cleanSegment(query).replace(/[-_]+/g, ' ').trim()
  if (!text) return null
  return `${POLYMARKET_BASE_URL}/search?query=${encodeURIComponent(text)}`
}

function buildKalshiSearchUrl(query: NullableString): string | null {
  const text = cleanSegment(query).replace(/[-_]+/g, ' ').trim()
  if (!text) return null
  return `${KALSHI_BASE_URL}/markets?search=${encodeURIComponent(text)}`
}

// Shared resolver for trader/order rows where payloads are often sparse and
// direct API URLs may be stale. Canonical builders are preferred first.
export function getTraderOrderPlatformLinks(input: TraderOrderLinkInput): OpportunityPlatformLinks {
  const payload = toRecord(input.payload)
  const marketId = cleanSegment(input.marketId)
  const marketQuestion = cleanSegment(input.marketQuestion)
  const payloadPlatform = cleanSegment(
    String(
      payload.platform
      || payload.exchange
      || payload.execution_platform
      || payload.market_platform
      || payload.source_platform
      || ''
    )
  ).toLowerCase()
  const sourcePlatform = normalizeSourcePlatform(input.source)
  const inferredPlatform = payloadPlatform === 'kalshi' || payloadPlatform === 'polymarket'
    ? (payloadPlatform as MarketPlatform)
    : inferMarketPlatform({
        platform: sourcePlatform || payloadPlatform || undefined,
        marketId,
        marketSlug: String(payload.market_slug || payload.slug || ''),
        conditionId: String(payload.condition_id || payload.conditionId || (isConditionId(marketId) ? marketId : '') || ''),
        eventTicker: String(payload.event_ticker || payload.eventTicker || ''),
      })

  const eventSlug = String(payload.event_slug || '').trim() || null
  const marketSlug = String(payload.market_slug || payload.slug || '').trim() || null
  const conditionId = String(payload.condition_id || payload.conditionId || (isConditionId(marketId) ? marketId : '') || '').trim() || null
  const marketTicker = String(payload.market_ticker || payload.kalshi_ticker || payload.ticker || '').trim() || null
  const eventTicker = String(payload.event_ticker || payload.eventTicker || '').trim() || null
  const seriesTicker = String(payload.series_ticker || payload.seriesTicker || '').trim() || null

  let polymarketUrl = buildPolymarketMarketUrl({
    eventSlug,
    marketSlug,
    marketId,
    conditionId,
  })
  let kalshiUrl = buildKalshiMarketUrl({
    marketTicker,
    eventTicker,
    eventSlug,
    seriesTicker,
  })

  if (!polymarketUrl && inferredPlatform !== 'kalshi') {
    polymarketUrl = buildPolymarketSearchUrl(marketQuestion || marketSlug || eventSlug)
  }
  if (!kalshiUrl && inferredPlatform === 'kalshi') {
    kalshiUrl = buildKalshiSearchUrl(marketQuestion || marketTicker || eventTicker)
  }

  const marketLinks: OpportunityLinkEntry[] = []
  if (polymarketUrl) marketLinks.push({ platform: 'polymarket', url: polymarketUrl })
  if (kalshiUrl) marketLinks.push({ platform: 'kalshi', url: kalshiUrl })
  return { polymarketUrl, kalshiUrl, marketLinks }
}
